### DISCUSSION TITLE
정적 팩토리 메소드의 장단점?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#29] [Clean Code]
.

**Comment by gyungchan-jo** [comment]
장점:
1. 이름을 가질 수 있음: 생성자와는 달리, 정적 팩터리 메서드는 이름을 가질 수 있습니다. 이를 통해 해당 메서드가 반환하는 객체의 의도나 용도를 명확하게 표현할 수 있습니다.
2. 반환 타입의 하위 타입 객체 반환 가능: 정적 팩터리 메서드는 생성될 객체의 하위 타입을 반환할 수 있습니다. 이를 활용하여 클라이언트에게 실제 생성된 객체의 타입을 숨길 수 있습니다.
3. 캐싱 가능: 정적 팩터리 메서드는 객체를 생성할 때 캐싱을 활용할 수 있습니다. 이를 통해 불필요한 객체 생성을 방지하고 성능을 향상시킬 수 있습니다.

단점:
1. 상속을 하려면 public 혹은 protected 생성자가 필요하니 정적 팩토리 메서드만 제공하면 하위 클래스를 만들 수 없다.
2. 정적 팩토리 메서드는 프로그래머가 찾기 어렵다. 
3. 정적 팩터리 메서드를 이용한 객체 생성 방식은 생성자를 사용하는 것보다 조금 더 직관적이지 않을 수 있습니다.

**Comment by le2sky** [comment]
정적 팩터리 메서드는 객체의 생성을 담당하는 클래스 메서드입니다.

정적 팩터리 메서드의 장점은 다음과 같습니다.

이름을 가질 수 있습니다.
인스턴스 캐싱이 가능합니다.
반환 타입의 하위 타입 객체를 반환할 수 있습니다.
부 생성자가 복잡할 경우, 정적 팩터리 메서드를 고려할 수 있습니다. 

단점은 다음과 같습니다.

정적 팩터리 메서드를 다른 개발자들이 찾기 어려울 수 있습니다.
생성자를 사용하지 않는 방식이 다른 개발자에게 어색하게 느껴질 수 있습니다.
남용할 시 생성자를 잊어버리게 될 수 있습니다. 생성하기에 복잡한 객체는 무언가 문제가 있을 확률이 높은데도 불구하고, 정적 팩터리 메서드로 잘못을 덮을 수도 있습니다.

**Comment by GIVEN53** [comment]
이름 부여
하위 타입 반환
캐싱

찾기 어려움
생성자 까먹음
